src/components/swap/SwapHeader.tsx

```diff
@@ -13,14 +22,136 @@ const StyledSwapHeader = styled.div`
   color: ${({ theme }) => theme.textSecondary};
 `
 
+const Dot = styled.span`
+  height: 8px;
+  width: 8px;
+  background-color: ${({ theme }) => theme.accentActive};
+  border-radius: 50%;
+  margin: 0px 4px;
+  display: inline-block;
+`
+
+const TextHeader = styled.div<{ color: string; marginLeft?: string; isClickable?: boolean; disabled?: boolean }>`
+  color: ${({ color }) => color};
+  opacity: ${({ disabled }) => disabled && '60%'};
+  margin-left: ${({ marginLeft }) => marginLeft};
+  margin-right: 8px;
+  display: flex;
+  flex-direction: row;
+  cursor: ${({ isClickable }) => isClickable && 'pointer'};
+  justify-content: center;
+  align-items: center;
+`
+
+const TOKEN_SAFETY_ARTICLE =
+  'https://support.uniswap.org/hc/en-us/articles/11306664890381-Why-isn-t-MoonPay-available-in-my-region-'
+
+// TODO(lynnshaoyu): add analytics and logging
+// TODO(lynnshaoyu): add animation & timeout to buy fiat unavailable toast
 export default function SwapHeader({ allowedSlippage }: { allowedSlippage: Percent }) {
+  const theme = useTheme()
+  const { account } = useWeb3React()
+  const openFiatOnRampModal = useOpenModal(ApplicationModal.FIAT_ONRAMP)
+  const fiatOnRampButtonEnabled = useFiatOnRampButtonEnabled()
+
+  const [buyFiatClicked, setBuyFiatClicked] = useBuyFiatClicked()
+  const [continueBuyFiatFlowPostRegionCheck, setContinueBuyFiatFlowPostRegionCheck] = useState(false)
+  const [continueBuyFiatFlowPostWalletModal, setContinueBuyFiatFlowPostWalletModal] = useState(false)
+
+  const [checkFiatRegionAvailability, setCheckFiatRegionAvailability] = useState(false)
+  const {
+    available: fiatOnrampAvailable,
+    availabilityChecked: fiatOnrampAvailabilityChecked,
+    error,
+    loading: fiatOnrampAvailabilityLoading,
+  } = useFiatOnrampAvailability(checkFiatRegionAvailability)
+
+  const disableBuyCryptoButton = Boolean(
+    error || (!fiatOnrampAvailable && fiatOnrampAvailabilityChecked) || fiatOnrampAvailabilityLoading
+  )
+
+  const [walletDrawerOpen, toggleWalletDrawer] = useWalletDrawer()
+
+  const handleBuyCrypto = useCallback(() => {
+    if (!fiatOnrampAvailabilityChecked) {
+      setCheckFiatRegionAvailability(true)
+      setContinueBuyFiatFlowPostRegionCheck(true)
+    } else if (fiatOnrampAvailable && !account && !walletDrawerOpen) {
+      toggleWalletDrawer()
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(true)
+    } else if (fiatOnrampAvailable && account) {
+      openFiatOnRampModal()
+      setBuyFiatClicked(true)
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(false)
+    } else if (!fiatOnrampAvailable) {
+      setBuyFiatClicked(true)
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(false)
+    }
+  }, [
```

Comment:
can you explain what the behavior here is doing 🤔 

is this essentially like incrementing steps in progress in the flow?

Reply:
yep exactly! it took me a few attempts to get this to reflect the flow we want without bugs XD 

so this is what is happening:
1. upon first attempt to click buy, make the region check (setCheckFiatRegionAvailability(true)) to see if fiat on ramps is available for the user.
2. if its UNAVAILABLE, we reach the 4th if statement case. this is the end of the flow. 
3. if it is AVAILABLE, we need to check if their account is connected. if not, open the wallet modal. however, we need to remember to open the fiat on ramp modal AFTER the wallet is connected. hence the continueBuyFiatFlowPostWalletModal state variable. 
4. if it is AVAILABLE and the account is now connected, we can finally show the fiat on ramp modal. this is also the end of the flow.

Reply:
flow makes sense to me, a few small things that stand out to me:
1) it's not obvious from this code what happens if the user already has the wallet drawer open, isn't connected, and clicks on 'buy' - are they sort of "stuck"? maybe this is more of a product question
2) what happens if a user clicks the button again in the middle of the flow? the button should probably be disabled if in the middle of the flow, i think
3) it's a bit hard to read and it's not immediately obvious how this callback + the useeffect below work together to increment state.

a few suggestions on how to improve readability:

1. add this github comment as a comment in code
2. it's a bit confusing that both the button click and the useEffect use this function. to me it feels like it's comingling the concern of starting the flow vs incrementing the flow, and to me it feels like it should be different because starting the flow is a user action vs incrementing the flow is automatic
3. this is just a suggestion so feel free to change up as you see fit or not use at all: the number of variables `checkFiatRegionAvailability`, `continueBuyFiatFlowPostWalletModal`, and `continueBuyFiatFlowPostRegionCheck` could probably be condensed. essentially, there's only one possible state for a user to be in at a time, so instead of setting true/false on a bunch of different individual state variables, could you use strings to represent the curr step instead of boolean combinations? 
4. you could probably break out the `setBuyFiatClicked` handling into its own useEffect that only updates once availability status has been checked, and then add a comment there about what `setBuyFiatClicked` is doing (e.g. hide blue dot if user already attempted the flow)

Reply:
also you could probably break out these hooks into a separate one called like `useFiatOnRampStep` and then add unit tests for it since it's a rather complicated flow

Reply:
I find modeling flows like this as a state machine helpful for readability

Reply:
1. Done
2.  I think re-using `handleBuyCrypto` is the most efficient way to code the desired behavior here, but understand the concern / how it can look confusing. To make it clearer, I added a comment above the second useEffect: `// Increment buy fiat flow state automatically, when requisite state changes have occured.` So the `handleBuyCrypto` is ALWAYS called when a user clicks on the Buy button, but is called conditionally when state is incremented automatically by the useEffect.
4. This makes sense, I condensed the state vars into the enum 'BuyFiatFlowState'
5. This makes sense too, I used another useEffect to handle "terminal" states of 'BuyFiatFlowState', which I defined as: `const TerminalBuyFiatFlowStates = new Set([
  BuyFiatFlowState.AVAILABLE_IN_REGION,
  BuyFiatFlowState.UNAVAILABLE_IN_REGION,
])`

Reply:
Also to answer the first 3 things that stand out: 

1. Good point, I disabled the Buy button now when the wallet drawer is open, and user is mid-flow.
2. Done.
3. I hope the comments, use of BuyFiatFlowState, and separation of concerns into multiple useEffects have made it clearer and easier to read 😁🤞

Reply:
Ok wait @tinaszheng I removed the other useEffect for this reason: https://github.com/Uniswap/interface/pull/6240/files#r1156534032